### PR TITLE
Add support for attention masks in BertEncoder

### PR DESCRIPTION
**Pull request by danieldk**
Normally, we want to mask attention for inactive timesteps, so that
their hidden states do not leak into the representations of active
time steps.

The API differs from Hugging Face transformers. In Hugging Face
transformers, the mask is converted to appropriate values that result
in near-zero probabilities in BertModel. This means that BertEncoder
already expects this special representation. However, since
BertEncoder (or even BertLayer) is the typical 'entry point' in
sticker-transformers, we perform the conversion in BertSelfAttention.

### CODE REVIEW

File: src/bert_model.rs [hunk 1]
```diff
@@ -654,6 +694,17 @@ mod tests {
         ]
     }
 
+    fn seqlen_to_mask(seq_lens: Tensor, max_len: i64) -> Tensor {
+        let batch_size = seq_lens.size()[0];
+        Tensor::arange(max_len, (Kind::Int, Device::Cpu))
```
Comment (twuebi) [inline src/bert_model.rs:692]: Does pinning the mask to CPU work when running on GPU?

Comment (danieldk) [reply src/bert_model.rs:692]: I don't know. Perhaps it would incur an extra transfer/copy each time? But this is only used for the unit test, which runs on the CPU anyway. If we need a crate-wide `seqlen_to_mask` at some point, I guess we should add the device as a parameter.

Comment (twuebi) [reply src/bert_model.rs:692]: Arf, my bad, for some reason I thought this was being used to create masks outside testing.

Comment (twuebi) [reply src/bert_model.rs:692]: Anyways, AFAIK, torch does not automatically copy tensors to the target device, in PyTorch things break if we pass a CPU tensor into a GPU model.

**Comment by twuebi** [comment]
>  However, since
> BertEncoder (or even BertLayer) is the typical 'entry point' in
> sticker-transformers, we perform the conversion in BertSelfAttention.

Moving the masking to `BertSelfAttention` means we perform the broadcasted multiplication per self-attention layer instead just once, what speaks against pre-computing the mask?

**Comment by danieldk** [comment]
> Moving the masking to `BertSelfAttention` means we perform the broadcasted multiplication per self-attention layer instead just once, what speaks against pre-computing the mask?

I couldn't make my mind up. This obviously has efficiency benefits, but it makes it easier to misuse the interface as long as `BertSelfAttention` is public. It's one of the things that is not so nice in the Python implementation, e.g. if you pass in a mask like that to `BertEncoder`, it will be wrong. So, if we compute it in `BertEncoder`, the other data structures should probably become private.

**Comment by twuebi** [comment]
Maybe we should just test the impact of recalculating the mask for every layer? Perhaps we could also clarify things by renaming the mask parameters to something that reflects the difference? Thinking of something like bool mask vs negative softmax mask. Maybe we could cache the mask such that it's not recomputed?

**Comment by danieldk** [comment]
> Maybe we should just test the impact of recalculating the mask for every layer? Perhaps we could also clarify things by renaming the mask parameters to something that reflects the difference? Thinking of something like bool mask vs negative softmax mask. Maybe we could cache the mask such that it's not recomputed?

We could use an enum to mark what it is. But for now, it would also be ok to do it in the encoder and just hide the other data structures (which may not be a bad idea anyway, since the parts may still change in their API).

**Comment by danieldk** [comment]
I think I have found an elegant solution. The conversion now happens in `BertEncoder`, but the lower-level layers accept the `LogitsMask` type, which can only be constructed from a `bool` tensor and is expected in the correct way.

**Review by twuebi — APPROVED** [review]
